Initialize Headers Dictionary Only Once

markekraus opened the pull request:
Switch the `WebResponseObject.Headers` Dictionary to initialize once instead of creating a new dictionary on every get. Moved logic to `WebResponseHelper.GetHeadersDictionary()` as this code will be reused outside of `WebResponseObject`.

Closes #3842 

This changes internal functionality only. Existing tests for the `Headers` property are sufficient.

msftclas commented:
@markekraus,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

markekraus committed:
[Feature] Create Headers Only Once and Only If Called

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseObject.CoreClr.cs:
```diff
@@ -31,23 +31,28 @@ public partial class WebResponseObject
             }
         }
 
+        private Dictionary<string, IEnumerable<string>> headers = null;
```
Should it be `_headers`? [naming-conventions](https://github.com/PowerShell/PowerShell/blob/ede580e7d19c1aecdb708ba9a09a7beaee5a0a14/docs/dev-process/coding-guidelines.md#naming-conventions)

markekraus replied:
Fixed, thanks.

markekraus committed:
{feature] headers -> _headers

markekraus committed:
[Feature] Move logic to WebResponseObject for reuse

This code has potential reuse in InvokeRestMethodCommand. Moving it to WebResponseHelper

markekraus commented:
I realized in planning for #4845 the headers dictionary logic would need to be available outside of `WebResponseObject` so I have moved it to `WebResponseHelper`.

iSazonov commented:
@markekraus Have we already tests? If so please add the comment in PR description. And about WebResponseHelper too.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:
```diff
@@ -16,6 +18,24 @@ internal static string GetCharacterSet(HttpResponseMessage response)
+            {
+                foreach (var entry in response.Content.Headers)
+                {
+                    headers[entry.Key] = entry.Value;
```
Interesting, can there be duplicate headers in `response.Headers` and `response.Content.Headers` ? Have second ones a high ptiority?

markekraus replied:
No. The headers in `Response.Content.Headers` will never exist in `response.Headers`. #4494 added this to address the lack of `Content-Type` in the `WebResponseObject.Headers`

iSazonov replied:
Please clarify - my understanding from #4494 is that `headers` already contains `response.Content.Headers`, correct?

markekraus replied:
That was add in #4494. Previous to that, `WebResponseObject.Headers` did not contain any of the headers in `HttpResponseMessage.Content.Headers`. CoreFX has split the response headers so that those related to content are always in `HttpResponseMessage.Content.Headers` and the rest of the response headers are in `HttpResponseMessage.Headers`. 

To clarify, I'm not adding or changing any of the logic in this PR for the creation of the Dictionary. Just moving it so that a new dictionary is not created on every Get to `Headers` and to make the logic available outside of `WebResponseObject`. The existing logic is there on purpose and we have existing tests to ensure it.

iSazonov replied:
Thanks for clarify.
So `foreach (var entry in response.Content.Headers) ` is duplicate code and could we remove it?
Sorry for late question - if CoreFX split the headers why we join them again? Only for backward compatibility?

markekraus replied:
We are joining them for backwards compat. In 5.1 and earlier all headers are in `WebResponseObject.Headers`. In 6.0 the content headers are buried in `WebResponseObject.BaseResponse.Content.Headers` unless we promote them to the `WebResponseObject.Headers` dictionary. One would reasonably expect the `Headers` property to contain all headers.

daxian-dbw replied:
@markekraus can you please add a comment to the code to clarify that there won't be duplicate headers in `response.Headers` and `response.Content.Headers`? It would be very helpful to other people who look at the code later.

markekraus replied:
@daxian-dbw should I also include a note about the content headers being added to the headers dictionary for backwards compatibility? or is the note about the distinct headers sufficient?

daxian-dbw replied:
I think it would also be helpful to mention the backward compatibility issue. Thanks @markekraus!

markekraus replied:
Comments added.

markekraus commented:
@iSazonov I have updated the PR description RE: Testing and move to `WebResponseHelper`.

markekraus committed:
Add comment to address PR Feedback.

daxian-dbw reviewed (approved):
LGTM. Thanks!

iSazonov commented:
@markekraus Thanks for the fix!
